content/post/clojure-ast.md

```diff
@@ -0,0 +1,162 @@
+            ana/var?         var?]
+    (with-env env
+      (ana/analyze form empty-env))))
+```
+
+So we call `analyze` with our custom parameters but use the functions for macro expansion and parsing from `clojure.tools.analyzer.jvm`. Easy, right?
+
+Well, we are not there yet. If your code uses macros you'll find them, obviously, expanded. But that's not a very nice AST representation. At least, it was not a very good one for us. So, instead of using the `macroexpand-1` function from the jvm analyzer, we use our own that doesn't actually do anything.
+
+```clojure
+(defn macroexpand-noexpand [form env] form)
+```
+
+Ok, now we're done, right? No we are not. If the code you're testing this with contains a `recur` form you will get an error because you can only `recur` from a tail position. What happens here is that macros are not expanded and, obviously, `loop` is a macro.
+
+We take a look at the code of the analyzer to see that it actually has a [`recur` parser where this behaviour is specifically checked](https://crossclj.info/ns/org.clojure/tools.analyzer/0.6.9/clojure.tools.analyzer.html#_parse-recur).
+
+What can we do now? Roll our own parse function with our own `recur` parser that omits these checks.
+
+```clojure
+(defn parse-recur
+  "Parse recur is the same parse function as the one in clojure.tools.analyzer
+  but it does not validate it."
+  [[_ & exprs :as form] env]
+  (let [exprs (mapv (ana/analyze-in-env (ctx env :ctx/expr)) exprs)]
+    {:op          :recur
+     :env         env
+     :form        form
+     :exprs       exprs
+     :children    [:exprs]}))
+
+(defn parse-forms
+  [form env]
+  ((case (first form)
+     monitor-enter        ana.jvm/parse-monitor-enter
+     monitor-exit         ana.jvm/parse-monitor-exit
+     clojure.core/import* ana.jvm/parse-import*
+     reify*               ana.jvm/parse-reify*
+     deftype*             ana.jvm/parse-deftype*
+     case*                ana.jvm/parse-case*
+     do                   ana/parse-do
+     if                   ana/parse-if
+     new                  ana/parse-new
+     quote                ana/parse-quote
+     set!                 ana/parse-set!
+     try                  ana/parse-try
+     throw                ana/parse-throw
+     def                  ana/parse-def
+     .                    ana/parse-dot
+     let*                 ana/parse-let*
+     letfn*               ana/parse-letfn*
+     loop*                ana/parse-loop*
+     recur                parse-recur
+     fn*                  ana/parse-fn*
+     var                  ana/parse-var
+     #_:else              ana/parse-invoke)
+   form env))
+```
+
+Now, if we glue it all toghether, we will get our AST for all the forms analyzed in isolation. Then, the AST of the program would be the list of the ASTs of all its forms.     
+
+You can take a look at the complete implementation of the bblfsh clojure driver [here](https://github.com/erizocosmico/clojure-driver/tree/feature/clj-parser).
```

Comment:
s/erizocosmico/bblfsh/
Yeah, I know it is still not there. But needs to before releasing the post.